File: app/src/main/java/com/github/abhijitpparate/keeps/data/auth/AuthSource.java
package com.github.abhijitpparate.keeps.data.auth;

import com.facebook.AccessToken;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.firebase.auth.FirebaseUser;
import com.twitter.sdk.android.core.TwitterSession;

import org.json.JSONObject;

import io.reactivex.Completable;
import io.reactivex.Maybe;

public interface AuthSource {

    Completable createNewAccount(Credentials credentials);

    Completable attemptLogin(Credentials credentials);

    Completable deleteUser();

    Completable logoutUser();

    Completable reAuthenticateUser(String password);

    void setReturnFail(boolean bool);
    void setAllowRegistration(boolean bool);

    Maybe<User> getUser();
    Maybe<User> attemptGoogleLogin(GoogleSignInAccount account);
    Maybe<User> attemptFacebookLogin(AccessToken token);
    Maybe<User> attemptTwitterLogin(TwitterSession session);
}

File: app/src/main/java/com/github/abhijitpparate/keeps/data/auth/AuthSource.kt
package com.github.abhijitpparate.keeps.data.auth

import com.facebook.AccessToken
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.google.firebase.auth.FirebaseUser
import com.twitter.sdk.android.core.TwitterSession

import org.json.JSONObject

import io.reactivex.Completable
import io.reactivex.Maybe

interface AuthSource {

    fun createNewAccount(credentials: Credentials): Completable

    fun attemptLogin(credentials: Credentials): Completable

    fun deleteUser(): Completable

    fun logoutUser(): Completable

    fun reAuthenticateUser(password: String): Completable

    fun setReturnFail(bool: Boolean)
    fun setAllowRegistration(bool: Boolean)

    val user: Maybe<User>
    fun attemptGoogleLogin(account: GoogleSignInAccount): Maybe<User>
    fun attemptFacebookLogin(token: AccessToken): Maybe<User>
    fun attemptTwitterLogin(session: TwitterSession): Maybe<User>
}
